File: backend/app/services/agent_tool_registry.py
"""
Agent Tool Registry
===================
The single source of truth for what the planner is allowed to emit.

The same specs generate the tool documentation inside the prompt AND validate
the model's reply, so a tool can never be described to the planner without
being validatable, or accepted without being described. Adding a tool is a
one-place change — the planner itself needs no edits (Step 18).
"""

from dataclasses import dataclass
from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple
from urllib.parse import urlparse

# Only ordinary web navigation. Mirrors the Browser Runtime's own allowlist so
# a blocked scheme is rejected before it ever reaches the browser.
ALLOWED_URL_SCHEMES = {"http", "https"}

MAX_TEXT_ARGUMENT_LENGTH = 5_000


class ToolValidationError(ValueError):
    """Raised when a planner reply does not satisfy the registry."""


@dataclass(frozen=True)
class ToolParameter:
    name: str
    kind: str  # "string" | "number" | "element_id" | "url"
    required: bool
    description: str


@dataclass(frozen=True)
class ToolSpec:
    name: str
    description: str
    parameters: Tuple[ToolParameter, ...] = ()
    """Which subsystem executes the tool. New categories need no planner change."""
    category: str = "browser"
    """When true the runtime must obtain explicit human consent before executing."""
    requires_approval: bool = False
    """True for tools the runtime resolves itself rather than sending to the browser."""
    handled_by_loop: bool = False

    def signature(self) -> str:
        """Compact one-line form used in the planner prompt."""
        if not self.parameters:
            return f'{{"tool": "{self.name}"}}'

        rendered = ", ".join(
            f'"{parameter.name}": <{parameter.kind}{"" if parameter.required else "?"}>'
            for parameter in self.parameters
        )
        return f'{{"tool": "{self.name}", "arguments": {{{rendered}}}}}'


TOOL_SPECS: Tuple[ToolSpec, ...] = (
    ToolSpec(
        name="click",
        description=(
            "Click an element on the page (button, link, option, tab). "
            "Never use click on a Search button before entering the search query into the search input."
        ),
        parameters=(
            ToolParameter("target", "element_id", True, "Element id from the page state, e.g. e4"),
        ),
    ),
    ToolSpec(
        name="fill",
        description=(
            "Replace the contents of an editable input field (role 'textbox', 'searchbox', 'combobox', 'input') "
            "with a text value. Always use this to enter search queries or form data."
        ),
        parameters=(
            ToolParameter("target", "element_id", True, "Element id of the input field"),
            ToolParameter("value", "string", True, "Text to enter"),
        ),
    ),
    ToolSpec(
        name="type",
        description="Type text into whatever currently has focus.",
        parameters=(
            ToolParameter("text", "string", True, "Text to type"),
        ),
    ),
    ToolSpec(
        name="press_key",
        # The scope note matters: key events are delivered to the page, not to
        # the browser, so a planner reaching for Control+t to open a tab
        # dispatches a keystroke that succeeds and does nothing.
        description=(
            "Press a single key inside the page, e.g. Enter, Tab, Escape, ArrowDown. "
            "Use 'Enter' after 'fill' to submit a search. "
            "Browser-level shortcuts such as Control+t or Control+w have no effect here "
            "— use open_tab to open a tab."
        ),
        parameters=(
            ToolParameter("key", "string", True, "Key name"),
        ),
    ),
    ToolSpec(
        name="navigate",
        description="Load a new http(s) url.",
        parameters=(
            ToolParameter("url", "url", True, "Absolute http or https url"),
        ),
    ),
    ToolSpec(
        name="open_tab",
        description=(
            "Open a new browser tab and switch to it, optionally loading a url. "
            "This is the only way to open a tab. To load a url in the tab already "
            "showing, use navigate instead."
        ),
        parameters=(
            ToolParameter("url", "url", False, "Absolute http or https url to open in the new tab"),
        ),
        # The tab strip is renderer state, so the agent loop opens the tab and
        # re-points itself at it; the Browser Runtime never sees this tool.
        handled_by_loop=True,
    ),
    ToolSpec(
        name="scroll",
        description="Scroll the page or an element. Positive deltaY scrolls down.",
        parameters=(
            ToolParameter("target", "element_id", False, "Element to scroll over"),
            ToolParameter("deltaX", "number", False, "Horizontal pixels"),
            ToolParameter("deltaY", "number", False, "Vertical pixels"),
        ),
    ),
    ToolSpec(
        name="upload",
        description="Attach files to a file input. Always asks the user to choose the files.",
        parameters=(
            ToolParameter("target", "element_id", True, "Element id of the file input"),
        ),
        category="files",
        requires_approval=True,
    ),
    ToolSpec(
        name="wait",
        description="Wait for the page to finish loading or settle.",
        parameters=(
            ToolParameter("timeoutMs", "number", False, "Maximum wait in milliseconds"),
        ),
    ),
    ToolSpec(
        name="extract",
        description="Record information from the current page into working memory.",
        parameters=(
            ToolParameter("note", "string", True, "What was found"),
        ),
        handled_by_loop=True,
    ),
    ToolSpec(
        name="finish",
        description="The goal is complete, or cannot be completed. Always end with this.",
        parameters=(
            ToolParameter("summary", "string", True, "Outcome for the user"),
        ),
        handled_by_loop=True,
    ),
)

# Mutable so tools can be added at import time by other modules without
# editing the planner. `register_tool` is the only supported way in.
_REGISTRY: Dict[str, ToolSpec] = {spec.name: spec for spec in TOOL_SPECS}


def register_tool(spec: ToolSpec) -> None:
    """
    Adds a tool to the registry.

    The planner needs no change: the tool is documented in the prompt and
    validated on the way back from the same spec (Step 18). The runtime must
    still know how to execute the new tool — a tool the registry advertises but
    nothing can run would produce plans that always fail.

    Deliberately not registered here: `terminal` and `email`. Handing shell
    execution or outbound mail to an agent whose input includes attacker-
    controlled page text converts a prompt injection into remote code
    execution or spam relay. The mechanism supports them; the judgement is
    that this application should not.
    """
    if spec.name in _REGISTRY:
        raise ValueError(f"Tool '{spec.name}' is already registered.")
    _REGISTRY[spec.name] = spec


def all_tools() -> Tuple[ToolSpec, ...]:
    return tuple(_REGISTRY.values())


def requires_approval(tool_name: str) -> bool:
    spec = _REGISTRY.get(tool_name)
    return bool(spec and spec.requires_approval)


class _ToolsByName(Mapping[str, ToolSpec]):
    """Read-only view so callers cannot mutate the registry behind its back."""

    def __getitem__(self, key: str) -> ToolSpec:
        return _REGISTRY[key]

    def __iter__(self):
        return iter(_REGISTRY)

    def __len__(self) -> int:
        return len(_REGISTRY)


TOOLS_BY_NAME: Mapping[str, ToolSpec] = _ToolsByName()


def render_tool_catalogue() -> str:
    """The tool documentation block injected into the planner prompt."""
    lines = []
    for spec in all_tools():
        suffix = " (the user must approve this before it runs)" if spec.requires_approval else ""
        lines.append(f"- {spec.name}: {spec.description}{suffix}")
        lines.append(f"  {spec.signature()}")
    return "\n".join(lines)


def _validate_number(name: str, value: Any) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ToolValidationError(f"Argument '{name}' must be a number.")
    return float(value)


def _validate_string(name: str, value: Any) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ToolValidationError(f"Argument '{name}' must be a non-empty string.")
    if len(value) > MAX_TEXT_ARGUMENT_LENGTH:
        raise ToolValidationError(f"Argument '{name}' exceeds {MAX_TEXT_ARGUMENT_LENGTH} characters.")
    return value


def _validate_url(name: str, value: Any) -> str:
    url = _validate_string(name, value)
    parsed = urlparse(url)
    if parsed.scheme.lower() not in ALLOWED_URL_SCHEMES:
        raise ToolValidationError(
            f"Argument '{name}' must be an http(s) url; got scheme '{parsed.scheme or 'none'}'."
        )
    if not parsed.netloc:
        raise ToolValidationError(f"Argument '{name}' must be an absolute url with a host.")
    return url


def _validate_element_id(name: str, value: Any, known_element_ids: Optional[Sequence[str]]) -> str:
    element_id = _validate_string(name, value)
    # A planner that invents an element id is hallucinating, and acting on a
    # made-up id would click whatever happens to occupy that slot. Reject it
    # here so the runtime never sees it.
    if known_element_ids is not None and element_id not in known_element_ids:
        raise ToolValidationError(
            f"Argument '{name}' references unknown element '{element_id}'. "
            "Only element ids present in the current page state may be used."
        )
    return element_id


def validate_tool_call(
    payload: Any,
    known_element_ids: Optional[Sequence[str]] = None,
) -> Tuple[str, Dict[str, Any]]:
    """
    Validate a raw planner tool call.

    Returns (tool_name, cleaned_arguments). Unknown arguments are rejected
    rather than dropped, since a planner passing an argument we ignore is a
    planner whose intent we have not actually understood.
    """
    if not isinstance(payload, dict):
        raise ToolValidationError("Planner output must be a JSON object.")

    tool_name = payload.get("tool")
    if not isinstance(tool_name, str) or tool_name not in TOOLS_BY_NAME:
        raise ToolValidationError(
            f"Unknown tool '{tool_name}'. Valid tools: {', '.join(TOOLS_BY_NAME)}."
        )

    spec = TOOLS_BY_NAME[tool_name]
    raw_arguments = payload.get("arguments", {})
    if raw_arguments is None:
        raw_arguments = {}
    if not isinstance(raw_arguments, dict):
        raise ToolValidationError(f"Arguments for '{tool_name}' must be a JSON object.")

    known_parameters = {parameter.name: parameter for parameter in spec.parameters}
    unexpected = sorted(set(raw_arguments) - set(known_parameters))
    if unexpected:
        raise ToolValidationError(
            f"Tool '{tool_name}' does not accept argument(s): {', '.join(unexpected)}."
        )

    cleaned: Dict[str, Any] = {}
    for parameter in spec.parameters:
        if parameter.name not in raw_arguments:
            if parameter.required:
                raise ToolValidationError(f"Tool '{tool_name}' requires argument '{parameter.name}'.")
            continue

        value = raw_arguments[parameter.name]
        if parameter.kind == "number":
            cleaned[parameter.name] = _validate_number(parameter.name, value)
        elif parameter.kind == "url":
            cleaned[parameter.name] = _validate_url(parameter.name, value)
        elif parameter.kind == "element_id":
            cleaned[parameter.name] = _validate_element_id(parameter.name, value, known_element_ids)
        else:
            cleaned[parameter.name] = _validate_string(parameter.name, value)

    return tool_name, cleaned


# A queue exists to save planning round trips on obvious sequences, not to let
# the planner run unsupervised. Longer queues mean more actions between
# security scans.
MAX_QUEUE_LENGTH = 4

# Tools whose element ids come from the state the planner was given.
_ELEMENT_DEPENDENT = {"click", "fill", "upload", "scroll"}


def validate_tool_queue(
    payloads: Any,
    known_element_ids: Optional[Sequence[str]] = None,
) -> List[Tuple[str, Dict[str, Any]]]:
    """
    Validates a short action queue.

    Beyond per-call validation, the queue as a whole must make sense:

    * Nothing may follow `finish` — the task is over.
    * Nothing that targets an element may follow a `navigate`. Element ids
      describe the page the planner was looking at; after a navigation they
      refer to nodes that no longer exist, and acting on them would mean
      clicking whatever happens to occupy that slot on the new page.
    * A tool needing approval must be alone, so the user is asked about a
      specific action rather than a batch they cannot inspect.
    """
    if not isinstance(payloads, list) or not payloads:
        raise ToolValidationError("Planner must return at least one tool call.")
    if len(payloads) > MAX_QUEUE_LENGTH:
        raise ToolValidationError(f"Planner returned more than {MAX_QUEUE_LENGTH} queued actions.")

    validated: List[Tuple[str, Dict[str, Any]]] = []
    navigated = False

    for index, payload in enumerate(payloads):
        tool_name, arguments = validate_tool_call(payload, known_element_ids)

        if index > 0 and validated[-1][0] == "finish":
            raise ToolValidationError("No action may follow 'finish'.")

        if navigated and tool_name in _ELEMENT_DEPENDENT:
            raise ToolValidationError(
                f"'{tool_name}' cannot be queued after 'navigate': element ids from the "
                "previous page are invalid. Plan the next page in a separate step."
            )

        if requires_approval(tool_name) and len(payloads) > 1:
            raise ToolValidationError(
                f"'{tool_name}' needs user approval and must be planned on its own, not queued."
            )

        if tool_name == "navigate":
            navigated = True

        validated.append((tool_name, arguments))

    return validated

File: backend-node/src/services/agentToolRegistry.ts
// Port of backend/app/services/agent_tool_registry.py — the single source of
// truth for what the planner is allowed to emit. Same specs generate the
// prompt documentation AND validate the model's reply (see render_tool_catalogue
// and validate_tool_call/validate_tool_queue). No LLM dependency, so this ports
// independently of the Phase 4 planner/gateway work.

const ALLOWED_URL_SCHEMES = new Set(['http:', 'https:'])
const MAX_TEXT_ARGUMENT_LENGTH = 5_000

export class ToolValidationError extends Error {}

export type ToolParameterKind = 'string' | 'number' | 'element_id' | 'url'

export interface ToolParameter {
  name: string
  kind: ToolParameterKind
  required: boolean
  description: string
}

export interface ToolSpec {
  name: string
  description: string
  parameters: readonly ToolParameter[]
  /** Which subsystem executes the tool. New categories need no planner change. */
  category: string
  /** When true the runtime must obtain explicit human consent before executing. */
  requiresApproval: boolean
  /** True for tools the runtime resolves itself rather than sending to the browser. */
  handledByLoop: boolean
}

function param(name: string, kind: ToolParameterKind, required: boolean, description: string): ToolParameter {
  return { name, kind, required, description }
}

function spec(partial: Partial<ToolSpec> & Pick<ToolSpec, 'name' | 'description'>): ToolSpec {
  return {
    parameters: [],
    category: 'browser',
    requiresApproval: false,
    handledByLoop: false,
    ...partial,
  }
}

/** Compact one-line planner-prompt form, e.g. `{"tool": "click", "arguments": {"target": <element_id>}}`. */
export function toolSignature(s: ToolSpec): string {
  if (s.parameters.length === 0) return `{"tool": "${s.name}"}`
  const rendered = s.parameters
    .map((p) => `"${p.name}": <${p.kind}${p.required ? '' : '?'}>`)
    .join(', ')
  return `{"tool": "${s.name}", "arguments": {${rendered}}}`
}

const TOOL_SPECS: readonly ToolSpec[] = [
  spec({
    name: 'click',
    description:
      'Click an element on the page (button, link, option, tab). ' +
      'Never use click on a Search button before entering the search query into the search input.',
    parameters: [param('target', 'element_id', true, 'Element id from the page state, e.g. e4')],
  }),
  spec({
    name: 'fill',
    description:
      "Replace the contents of an editable input field (role 'textbox', 'searchbox', 'combobox', 'input') " +
      'with a text value. Always use this to enter search queries, form data, or filter/setting values ' +
      "(e.g. a price, quantity, or date field) — prefer it over 'type' whenever the field has an element id.",
    parameters: [
      param('target', 'element_id', true, 'Element id of the input field'),
      param('value', 'string', true, 'Text to enter'),
    ],
  }),
  spec({
    name: 'type',
    description: 'Type text into whatever currently has focus.',
    parameters: [param('text', 'string', true, 'Text to type')],
  }),
  spec({
    name: 'press_key',
    description:
      'Press a single key inside the page, e.g. Enter, Tab, Escape, ArrowDown. ' +
      "Use 'Enter' after 'fill' to submit a search. " +
      'Browser-level shortcuts such as Control+t or Control+w have no effect here ' +
      '— use open_tab to open a tab.',
    parameters: [param('key', 'string', true, 'Key name')],
  }),
  spec({
    name: 'navigate',
    description: 'Load a new http(s) url.',
    parameters: [param('url', 'url', true, 'Absolute http or https url')],
  }),
  spec({
    name: 'open_tab',
    description:
      'Open a new browser tab and switch to it, optionally loading a url. ' +
      'This is the only way to open a tab. To load a url in the tab already ' +
      'showing, use navigate instead.',
    parameters: [param('url', 'url', false, 'Absolute http or https url to open in the new tab')],
    handledByLoop: true,
  }),
  spec({
    name: 'scroll',
    description: 'Scroll the page or an element. Positive deltaY scrolls down.',
    parameters: [
      param('target', 'element_id', false, 'Element to scroll over'),
      param('deltaX', 'number', false, 'Horizontal pixels'),
      param('deltaY', 'number', false, 'Vertical pixels'),
    ],
  }),
  spec({
    name: 'upload',
    description: 'Attach files to a file input. Always asks the user to choose the files.',
    parameters: [param('target', 'element_id', true, 'Element id of the file input')],
    category: 'files',
    requiresApproval: true,
  }),
  spec({
    name: 'wait',
    description:
      'Wait for the page to finish loading or settle. Rarely needed: click/fill/type/press_key ' +
      'already wait for the page to settle before returning, so do not queue this right after one ' +
      'of them — only use it when something is loading that none of your actions triggered.',
    parameters: [param('timeoutMs', 'number', false, 'Maximum wait in milliseconds')],
  }),
  spec({
    name: 'extract',
    description:
      'Record information from the current page into working memory. The note must describe what ' +
      'is actually visible in the current page state (a field\'s value, the url, the results shown) ' +
      '— not what an earlier action was intended to accomplish.',
    parameters: [param('note', 'string', true, 'What was found')],
    handledByLoop: true,
  }),
  spec({
    name: 'finish',
    description: 'The goal is complete, or cannot be completed. Always end with this.',
    parameters: [param('summary', 'string', true, 'Outcome for the user')],
    handledByLoop: true,
  }),
]

// Mutable so tools can be added at runtime without editing the planner.
// `registerTool` is the only supported way in.
const REGISTRY = new Map<string, ToolSpec>(TOOL_SPECS.map((s) => [s.name, s]))

/**
 * Adds a tool to the registry. The planner needs no change: the tool is
 * documented in the prompt and validated on the way back from the same spec.
 * The runtime must still know how to execute the new tool.
 *
 * Deliberately not registered here: `terminal` and `email`. Handing shell
 * execution or outbound mail to an agent whose input includes attacker-
 * controlled page text converts a prompt injection into remote code
 * execution or spam relay. The mechanism supports them; the judgement is
 * that this application should not.
 */
export function registerTool(s: ToolSpec): void {
  if (REGISTRY.has(s.name)) {
    throw new Error(`Tool '${s.name}' is already registered.`)
  }
  REGISTRY.set(s.name, s)
}

/** Test-only escape hatch mirroring the Python tests' direct `_REGISTRY.pop(...)`. */
export function unregisterTool(name: string): void {
  REGISTRY.delete(name)
}

export function allTools(): readonly ToolSpec[] {
  return [...REGISTRY.values()]
}

export function requiresApproval(toolName: string): boolean {
  return REGISTRY.get(toolName)?.requiresApproval ?? false
}

export const TOOLS_BY_NAME: ReadonlyMap<string, ToolSpec> = REGISTRY

/** The tool documentation block injected into the planner prompt. */
export function renderToolCatalogue(): string {
  const lines: string[] = []
  for (const s of allTools()) {
    const suffix = s.requiresApproval ? ' (the user must approve this before it runs)' : ''
    lines.push(`- ${s.name}: ${s.description}${suffix}`)
    lines.push(`  ${toolSignature(s)}`)
  }
  return lines.join('\n')
}

function validateNumber(name: string, value: unknown): number {
  if (typeof value === 'boolean' || typeof value !== 'number' || Number.isNaN(value)) {
    throw new ToolValidationError(`Argument '${name}' must be a number.`)
  }
  return value
}

function validateString(name: string, value: unknown): string {
  if (typeof value !== 'string' || !value.trim()) {
    throw new ToolValidationError(`Argument '${name}' must be a non-empty string.`)
  }
  if (value.length > MAX_TEXT_ARGUMENT_LENGTH) {
    throw new ToolValidationError(`Argument '${name}' exceeds ${MAX_TEXT_ARGUMENT_LENGTH} characters.`)
  }
  return value
}

function validateUrl(name: string, value: unknown): string {
  const url = validateString(name, value)
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    throw new ToolValidationError(`Argument '${name}' must be an absolute url with a host.`)
  }
  if (!ALLOWED_URL_SCHEMES.has(parsed.protocol)) {
    throw new ToolValidationError(
      `Argument '${name}' must be an http(s) url; got scheme '${parsed.protocol.replace(':', '') || 'none'}'.`,
    )
  }
  if (!parsed.host) {
    throw new ToolValidationError(`Argument '${name}' must be an absolute url with a host.`)
  }
  return url
}

function validateElementId(name: string, value: unknown, knownElementIds?: readonly string[]): string {
  const elementId = validateString(name, value)
  // A planner that invents an element id is hallucinating, and acting on a
  // made-up id would click whatever happens to occupy that slot.
  if (knownElementIds !== undefined && !knownElementIds.includes(elementId)) {
    throw new ToolValidationError(
      `Argument '${name}' references unknown element '${elementId}'. ` +
        'Only element ids present in the current page state may be used.',
    )
  }
  return elementId
}

/**
 * Validate a raw planner tool call. Returns [toolName, cleanedArguments].
 * Unknown arguments are rejected rather than dropped.
 */
export function validateToolCall(
  payload: unknown,
  knownElementIds?: readonly string[],
): [string, Record<string, unknown>] {
  if (typeof payload !== 'object' || payload === null || Array.isArray(payload)) {
    throw new ToolValidationError('Planner output must be a JSON object.')
  }
  const record = payload as Record<string, unknown>

  const toolName = record.tool
  if (typeof toolName !== 'string' || !TOOLS_BY_NAME.has(toolName)) {
    throw new ToolValidationError(
      `Unknown tool '${String(toolName)}'. Valid tools: ${[...TOOLS_BY_NAME.keys()].join(', ')}.`,
    )
  }

  const toolSpec = TOOLS_BY_NAME.get(toolName)!
  let rawArguments = record.arguments
  if (rawArguments === null || rawArguments === undefined) rawArguments = {}
  if (typeof rawArguments !== 'object' || Array.isArray(rawArguments)) {
    throw new ToolValidationError(`Arguments for '${toolName}' must be a JSON object.`)
  }
  const rawArgs = rawArguments as Record<string, unknown>

  const knownParameterNames = new Set(toolSpec.parameters.map((p) => p.name))
  const unexpected = Object.keys(rawArgs).filter((k) => !knownParameterNames.has(k)).sort()
  if (unexpected.length > 0) {
    throw new ToolValidationError(`Tool '${toolName}' does not accept argument(s): ${unexpected.join(', ')}.`)
  }

  const cleaned: Record<string, unknown> = {}
  for (const p of toolSpec.parameters) {
    if (!(p.name in rawArgs)) {
      if (p.required) {
        throw new ToolValidationError(`Tool '${toolName}' requires argument '${p.name}'.`)
      }
      continue
    }
    const value = rawArgs[p.name]
    if (p.kind === 'number') cleaned[p.name] = validateNumber(p.name, value)
    else if (p.kind === 'url') cleaned[p.name] = validateUrl(p.name, value)
    else if (p.kind === 'element_id') cleaned[p.name] = validateElementId(p.name, value, knownElementIds)
    else cleaned[p.name] = validateString(p.name, value)
  }

  return [toolName, cleaned]
}

// A queue exists to save planning round trips on obvious sequences, not to let
// the planner run unsupervised. Longer queues mean more actions between
// security scans.
export const MAX_QUEUE_LENGTH = 4

// Tools whose element ids come from the state the planner was given.
const ELEMENT_DEPENDENT = new Set(['click', 'fill', 'upload', 'scroll'])

/**
 * Validates a short action queue. Beyond per-call validation, the queue as a
 * whole must make sense:
 *  - Nothing may follow `finish`.
 *  - Nothing that targets an element may follow a `navigate` (element ids from
 *    the previous page are invalid on the new one).
 *  - A tool needing approval must be alone in the queue.
 */
export function validateToolQueue(
  payloads: unknown,
  knownElementIds?: readonly string[],
): Array<[string, Record<string, unknown>]> {
  if (!Array.isArray(payloads) || payloads.length === 0) {
    throw new ToolValidationError('Planner must return at least one tool call.')
  }
  if (payloads.length > MAX_QUEUE_LENGTH) {
    throw new ToolValidationError(`Planner returned more than ${MAX_QUEUE_LENGTH} queued actions.`)
  }

  const validated: Array<[string, Record<string, unknown>]> = []
  let navigated = false

  for (let index = 0; index < payloads.length; index++) {
    const [toolName, args] = validateToolCall(payloads[index], knownElementIds)

    if (index > 0 && validated[index - 1]![0] === 'finish') {
      throw new ToolValidationError("No action may follow 'finish'.")
    }

    if (navigated && ELEMENT_DEPENDENT.has(toolName)) {
      throw new ToolValidationError(
        `'${toolName}' cannot be queued after 'navigate': element ids from the ` +
          'previous page are invalid. Plan the next page in a separate step.',
      )
    }

    if (requiresApproval(toolName) && payloads.length > 1) {
      throw new ToolValidationError(`'${toolName}' needs user approval and must be planned on its own, not queued.`)
    }

    if (toolName === 'navigate') navigated = true

    validated.push([toolName, args])
  }

  return validated
}
